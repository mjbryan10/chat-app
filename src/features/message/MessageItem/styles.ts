import styled, { ThemeColor, ThemeProps, Theme } from 'styled-components';

interface ContainerProps extends ThemeProps<Theme> {
   color: ThemeColor;
   isChained: boolean;
   isOwner: boolean;
}
interface HeaderProps extends ThemeProps<Theme> {
   color: ThemeColor;
}

const colorOrOwner = ({ theme, isOwner, color }: ContainerProps) => {
   return (isOwner ? theme.color.owner : theme.color[color]) ?? 'inherit';
};

const colorProp = ({ theme, color }: HeaderProps) => {
   return theme.color[color] ?? 'inherit';
};

export const Container = styled.div<ContainerProps>`
   background: ${({ theme }: ContainerProps) => theme.background.message ?? 'inherit'};
   border: ${colorOrOwner} 1.5px solid;
   min-width: 30%;
   max-width: 800px;
   margin: 5px;
   ${({ isOwner }) =>
      isOwner ? `margin: 5px 5px 5px auto;` : `margin: 5px auto 5px 5px;`}
   &::after {
      display: block;
      content: ' ';
      width: 100% + 3px;
      height: 10px;
      clip-path: polygon(100% 0, 0 100%, 0 0);
      background: ${colorOrOwner};
      margin: 0 -1.5px -10px -1.5px;
      ${({ isOwner }) => (isOwner ? `transform: rotateY(180deg); margin-right: -1.5px;` : null)}
      ${({ isChained }) => (isChained ? 'display: none;' : null)}
   }
`;

export const Header = styled.div<HeaderProps>`
   color: ${colorProp};
   display: flex;
   flex-flow: row nowrap;
   justify-content: space-between;
   align-items: center;
   padding: 0 0.5rem;
`;

export const Detail = styled.span`
   &:first-of-type {
      margin-right: auto;
   }
   &:last-of-type {
      margin-left: auto;
   }
`;

export const Body = styled.div<ThemeProps<Theme>>`
   padding: 0.5em;
   color: ${({theme}) => theme.color.text ?? 'auto'}
`;

export const BodyText = styled.p`
   color: ${({ theme }) => theme.color.text ?? 'inherit'};
`;
