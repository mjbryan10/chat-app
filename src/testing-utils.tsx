import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }