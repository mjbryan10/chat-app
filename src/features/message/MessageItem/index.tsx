import React, { FC } from 'react';
import * as S from './styles';
import { ThemeColor } from 'styled-components';
// import UserApi from 'Api/UserApi';
export interface Props {
   messageText: string;
   username: string;
   isChainMessage?: boolean;
   color: ThemeColor;
   time: string;
}

/**
 * A React Function Component responsible for rendering a message from the chat API.
 * @param message A message object from the API
 * @param isChaind A `boolean` to indicate if the message is part of a chain
 */
const MessageItem: FC<Props> = ({
   messageText,
   username,
   isChainMessage = false,
   color,
   time,
}) => {
   return (
      <S.Container color={color} isChained={isChainMessage}>
         <S.Header>
            <S.Detail>{username}</S.Detail>
            <S.Detail>{time}</S.Detail>
         </S.Header>
         <S.Body>{messageText}</S.Body>
      </S.Container>
   );
};

export default MessageItem;
