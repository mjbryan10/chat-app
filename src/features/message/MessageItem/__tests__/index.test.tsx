import React from 'react';
import { render, RenderResult } from 'testing-utils';
import MessageItem from '..';
import { cleanup } from '@testing-library/react';

jest.mock('axios');

describe('MessageItem Component', () => {
   const testMessage = {
      id: 1,
      senderId: 5,
      message: 'The first message',
      timestamp: '2020-09-04 08:15:00',
      conversationid: 2,
      status: 0,
   };
   let utils: RenderResult;
   describe('MessageItem intial render', () => {
      beforeEach(() => {
         utils = render(
            <MessageItem
               messageText={testMessage.message}
               username="Wessel"
               color="owner"
               time="08:15"
            />
         );
      });
      afterEach(() => {
         jest.resetAllMocks();
         cleanup();
      });

      test('should display message body', () => {
         const { getByText } = utils;

         expect(getByText(/The first message/i)).toBeVisible();
      });

      test('should display message sent time (MM:SS format)', () => {
         const { getByText } = utils;

         expect(getByText(/08:15/i)).toBeVisible();
      });

      test('should display username', () => {
         const { getByText } = utils;

         expect(getByText(/'Wessel'/i)).toBeVisible();
      });
   });
});
