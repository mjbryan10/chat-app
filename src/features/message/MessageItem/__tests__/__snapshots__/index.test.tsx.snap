// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageItem Component MessageItem intial render should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm ljuivB"
    color="red"
  >
    <div
      class="sc-AxirZ iCpfoH"
      color="red"
    >
      <span
        class="sc-AxiKw BLeyf"
      >
        You
      </span>
      <span
        class="sc-AxiKw BLeyf"
      >
        10:15
      </span>
    </div>
    <div
      class="sc-AxhCb fpilma"
    >
      The first message
    </div>
  </div>
</DocumentFragment>
`;
